File: app/src/main/java/com/mohammadkk/myaudioplayer/service/MediaService.java
package com.mohammadkk.myaudioplayer.service;

import android.app.Notification;
import android.app.PendingIntent;
import android.app.Service;
import android.content.ContentUris;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.media.MediaMetadata;
import android.media.MediaPlayer;
import android.os.Binder;
import android.os.Build;
import android.os.IBinder;
import android.support.v4.media.MediaMetadataCompat;
import android.support.v4.media.session.MediaSessionCompat;

import androidx.annotation.Nullable;
import androidx.core.app.NotificationCompat;

import com.mohammadkk.myaudioplayer.NotificationReceiver;
import com.mohammadkk.myaudioplayer.PlayerActivity;
import com.mohammadkk.myaudioplayer.R;
import com.mohammadkk.myaudioplayer.helper.MusicUtil;
import com.mohammadkk.myaudioplayer.model.Songs;

import java.io.IOException;
import java.util.ArrayList;
import static com.mohammadkk.myaudioplayer.MainActivity.songsAllList;
import static com.mohammadkk.myaudioplayer.MediaApplication.ACTION_NEXT;
import static com.mohammadkk.myaudioplayer.MediaApplication.ACTION_PLAY;
import static com.mohammadkk.myaudioplayer.MediaApplication.ACTION_PREVIOUS;
import static com.mohammadkk.myaudioplayer.MediaApplication.MUSIC_CHANNEL;

public class MediaService extends Service {

    IBinder bindService = new BindService();
    public static boolean isService = false;
    private static int servicePosition = 0;
    private ArrayList<Songs> mediaList = new ArrayList<>();
    private MediaPlayer mediaPlayer;
    private MediaSessionCompat mediaSessionCompat;
    private CallBackService callBackService;

    @Override
    public void onCreate() {
        super.onCreate();
        mediaList = songsAllList;
        mediaPlayer = new MediaPlayer();
        mediaSessionCompat = new MediaSessionCompat(getBaseContext(),"My Media Player");
        isService = true;
    }

    public ArrayList<Songs> getMediaList() {
        return mediaList;
    }
    public void setMediaList(ArrayList<Songs> mediaList) {
        this.mediaList = mediaList;
    }
    @Override
    public void onDestroy() {
        super.onDestroy();
        mediaPlayer.stop();
        isService = false;
    }
    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        String actionName = intent.getStringExtra("actionName");
        mediaPlayer.setOnCompletionListener(mp -> callBackService.setNextMusic());
        if (actionName != null) {
            switch (actionName) {
                case "playPause":
                    callBackService.setPlayPauseMusic();
                    break;
                case "next":
                    callBackService.setNextMusic();
                    break;
                case "previous":
                    callBackService.setPrevMusic();
                    break;
                case "stopService":
                    stopSelf();
                    break;
            }
        }
        return START_STICKY;
    }
    public void setCallBackService(CallBackService callBackService) {
        this.callBackService = callBackService;
    }
    public void reset() {
        mediaPlayer.reset();
    }
    public void createMediaPlayer(int position) throws IOException {
        mediaPlayer.setDataSource(this, MusicUtil.getUriPath(mediaList.get(position)));
        servicePosition = position;
        MediaMetadataCompat builder = new MediaMetadataCompat.Builder()
                .putString(MediaMetadata.METADATA_KEY_TITLE, mediaList.get(servicePosition).getTitle())
                .putString(MediaMetadata.METADATA_KEY_ARTIST, mediaList.get(servicePosition).getArtist())
                .build();
        mediaSessionCompat.setMetadata(builder);
    }
    public int getServicePosition() {
        return servicePosition;
    }
    public void prepare() throws IOException {
        mediaPlayer.prepare();
    }
    public void start() {
        mediaPlayer.start();
    }
    public boolean isPlaying() {
        return mediaPlayer.isPlaying();
    }
    public int getCurrentPosition() {
        return mediaPlayer.getCurrentPosition();
    }
    public int getDuration() {
        return mediaPlayer.getDuration();
    }
    public void seekTo(int mSec) {
        mediaPlayer.seekTo(mSec);
    }
    public void pause() {
        mediaPlayer.pause();
    }
    public void showNotification(int playImg) {
        Intent intent = new Intent(this, PlayerActivity.class);
        PendingIntent contentIntent = PendingIntent.getActivity(this, 0, intent, 0);
        Intent prevIntent = new Intent(this, NotificationReceiver.class).setAction(ACTION_PREVIOUS);
        PendingIntent prevPending = PendingIntent.getBroadcast(this, 0, prevIntent, PendingIntent.FLAG_UPDATE_CURRENT);
        Intent pauseIntent = new Intent(this, NotificationReceiver.class).setAction(ACTION_PLAY);
        PendingIntent pausePending = PendingIntent.getBroadcast(this, 0, pauseIntent, PendingIntent.FLAG_UPDATE_CURRENT);
        Intent nextIntent = new Intent(this, NotificationReceiver.class).setAction(ACTION_NEXT);
        PendingIntent nextPending = PendingIntent.getBroadcast(this, 0, nextIntent, PendingIntent.FLAG_UPDATE_CURRENT);
        Bitmap thumbInit = MusicUtil.getSongCover(getBaseContext(), MusicUtil.getUriPath(mediaList.get(servicePosition)));
        Bitmap thumb = thumbInit != null ? thumbInit : BitmapFactory.decodeResource(getResources(), R.drawable.ic_music_large);
        SharedPreferences pref = getSharedPreferences("cache_service", Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = pref.edit();
        editor.putInt("play_state", playImg);
        editor.putInt("play_pos", servicePosition);
        editor.apply();
        Notification notification =  new NotificationCompat.Builder(this, MUSIC_CHANNEL)
                .setSmallIcon(R.drawable.ic_songs)
                .setLargeIcon(thumb)
                .setContentTitle(mediaList.get(servicePosition).getTitle())
                .setContentText(mediaList.get(servicePosition).getArtist())
                .setContentIntent(contentIntent)
                .addAction(R.drawable.ic_skip_prev, "Previous", prevPending)
                .addAction(playImg, "Pause", pausePending)
                .addAction(R.drawable.ic_skip_next, "Next", nextPending)
                .setStyle(new androidx.media.app.NotificationCompat.MediaStyle().setMediaSession(mediaSessionCompat.getSessionToken()).setShowActionsInCompactView(0, 1, 2))
                .setPriority(NotificationCompat.PRIORITY_HIGH)
                .setOnlyAlertOnce(true)
                .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                .build();
        startForeground(1, notification);
    }
    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return bindService;
    }
    public class BindService extends Binder {
        public MediaService getService() {
            return MediaService.this;
        }
    }
}

File: app/src/main/java/com/mohammadkk/myaudioplayer/service/MediaService.kt
package com.mohammadkk.myaudioplayer.service

import android.app.PendingIntent
import android.app.Service
import android.content.Intent
import android.graphics.BitmapFactory
import android.media.AudioAttributes
import android.media.AudioManager
import android.media.MediaPlayer
import android.os.Binder
import android.os.IBinder
import android.support.v4.media.MediaMetadataCompat
import android.support.v4.media.session.MediaSessionCompat
import android.util.Log
import androidx.core.app.NotificationCompat
import com.mohammadkk.myaudioplayer.AudioApp
import com.mohammadkk.myaudioplayer.PlayerActivity
import com.mohammadkk.myaudioplayer.R
import com.mohammadkk.myaudioplayer.buildCacheApp
import com.mohammadkk.myaudioplayer.extension.getCoverTrack
import com.mohammadkk.myaudioplayer.extension.toContentUri
import com.mohammadkk.myaudioplayer.fragment.NowPlayerFragment
import com.mohammadkk.myaudioplayer.helper.BuildUtil
import com.mohammadkk.myaudioplayer.helper.Constants
import com.mohammadkk.myaudioplayer.model.Track
import java.io.IOException


class MediaService : Service(), MediaPlayer.OnCompletionListener, MediaPlayer.OnPreparedListener, MediaPlayer.OnErrorListener {
    private var bindService: IBinder = BindService()
    private var mediaPlayer: MediaPlayer? = null
    private var mMediaSession: MediaSessionCompat? = null
    private var listener: ServiceListener? = null
    private var serviceIndex = 0

    override fun onBind(intent: Intent): IBinder {
        return bindService
    }
    override fun onCreate() {
        super.onCreate()
        mediaPlayer = MediaPlayer()
        mMediaSession = MediaSessionCompat(baseContext, "MediaService")
        isExists = true
    }
    override fun onDestroy() {
        super.onDestroy()
        mediaPlayer?.stop()
        mediaPlayer?.release()
        mMediaSession?.isActive = false
        isExists = false
        buildCacheApp.requirClear()
    }
    fun setOnListenerService(listener: ServiceListener) {
        this.listener = listener
    }
    @Throws(IOException::class)
    fun createMediaPlayer(position: Int) {
        if (position >= 0 && position < trackList.size) {
            serviceIndex = position
            buildCacheApp.globalTrackIndexCaller = serviceIndex
            val track = trackList[position]
            mediaPlayer?.setDataSource(baseContext, track.id.toContentUri())
            mMediaSession?.isActive = true
            val metadata = MediaMetadataCompat.Builder()
                .putString(MediaMetadataCompat.METADATA_KEY_ALBUM, track.album)
                .putString(MediaMetadataCompat.METADATA_KEY_ARTIST, track.artist)
                .putString(MediaMetadataCompat.METADATA_KEY_TITLE, track.title)
                .putString(MediaMetadataCompat.METADATA_KEY_MEDIA_ID, track.id.toString())
                .putLong(MediaMetadataCompat.METADATA_KEY_DURATION, track.duration.toLong())
                .build()
            mMediaSession?.setMetadata(metadata)
        }
    }
    override fun onCompletion(mp: MediaPlayer?) {
        listener?.onNextTrack()
    }
    override fun onPrepared(mp: MediaPlayer?) {
        if (mediaPlayer?.isPlaying == false)
            mediaPlayer?.start()
    }
    override fun onError(mp: MediaPlayer?, what: Int, extra: Int): Boolean {
        val errorList = intArrayOf(
            MediaPlayer.MEDIA_ERROR_NOT_VALID_FOR_PROGRESSIVE_PLAYBACK,
            MediaPlayer.MEDIA_ERROR_UNSUPPORTED,
            MediaPlayer.MEDIA_ERROR_UNKNOWN,
            MediaPlayer.MEDIA_ERROR_SERVER_DIED,
            MediaPlayer.MEDIA_ERROR_IO
        )
        for (error in errorList) {
            if (what == error) {
                Log.i("error_media_player", "error code: $error, extra: $extra")
            }
        }
        return false
    }
    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        serviceIndex = buildCacheApp.globalTrackIndexCaller
        val tracks = buildCacheApp.getStoreTracks()
        if (tracks != null) trackList = tracks
        if (serviceIndex == -1 && trackList.isEmpty()) {
            stopSelf()
        }
        val actionName = intent?.getStringExtra("actionName")

        when (actionName ?: "") {
            "playPause" -> listener?.onPlayPauseTrack()
            "next" -> listener?.onNextTrack()
            "previous" -> listener?.onPreviousTrack()
            "stopService" -> {
                stopForeground(true)
                stopSelf()
            }
        }
        mediaPlayer?.setOnCompletionListener(this)
        mediaPlayer?.setOnPreparedListener(this)
        mediaPlayer?.setOnErrorListener(this)
        return START_STICKY
    }
    fun reset() {
        mediaPlayer?.reset()
    }
    fun setDefaultAudioStream() {
        if (BuildUtil.isOreoPlus()) {
            mediaPlayer?.setAudioAttributes(audioAttributes)
        } else {
            @Suppress("DEPRECATION")
            mediaPlayer?.setAudioStreamType(AudioManager.STREAM_MUSIC)
        }
    }
    @Throws(IOException::class)
    fun prepareAsync() {
        mediaPlayer?.prepareAsync()
    }
    fun start() {
        mediaPlayer?.start()
    }
    fun seekTo(mSec: Int) {
        mediaPlayer?.seekTo(mSec)
    }
    fun pause() {
        mediaPlayer?.pause()
    }
    val isPlaying: Boolean
        get() = mediaPlayer!!.isPlaying
    val currentPosition: Int
        get() = mediaPlayer!!.currentPosition
    val duration: Int
        get() = mediaPlayer!!.duration
    private val audioAttributes = AudioAttributes.Builder()
        .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
        .setUsage(AudioAttributes.USAGE_MEDIA)
        .build()
    fun showNotification(playImg: Int) {
        val flags = Constants.PENDING_INTENT_FLAG
        val mi = Intent(this, PlayerActivity::class.java)
        val contentIntent = PendingIntent.getActivity(this, 0, mi, flags)
        val prevIntent = Intent(this, NotificationReceiver::class.java).setAction(AudioApp.ACTION_PREVIOUS)
        val prevPending = PendingIntent.getBroadcast(this, 0, prevIntent, flags)
        val pauseIntent = Intent(this, NotificationReceiver::class.java).setAction(AudioApp.ACTION_PLAY)
        val pausePending = PendingIntent.getBroadcast(this, 0, pauseIntent, flags)
        val nextIntent = Intent(this, NotificationReceiver::class.java).setAction(AudioApp.ACTION_NEXT)
        val nextPending = PendingIntent.getBroadcast(this, 0, nextIntent, flags)
        val track = trackList[serviceIndex]
        var cover = getCoverTrack(track.id.toContentUri())
        cover = cover ?: BitmapFactory.decodeResource(resources, R.drawable.ic_music_large)
        buildCacheApp.iconPausedCaller = playImg
        buildCacheApp.titleTextCaller = track.title
        buildCacheApp.artistTextCaller = track.artist
        NowPlayerFragment.listener?.updateNowPlay(playImg, track)
        val notification = NotificationCompat.Builder(this, AudioApp.AUDIO_CHANNEL)
            .setSmallIcon(R.drawable.ic_track)
            .setLargeIcon(cover)
            .setContentTitle(track.title)
            .setContentText(track.artist)
            .setContentIntent(contentIntent)
            .addAction(R.drawable.ic_skip_previous, "Previous", prevPending)
            .addAction(playImg, "Pause", pausePending)
            .addAction(R.drawable.ic_skip_next, "Next", nextPending)
            .setStyle(
                androidx.media.app.NotificationCompat.MediaStyle()
                    .setShowActionsInCompactView(0, 1, 2)
                    .setMediaSession(mMediaSession?.sessionToken)
            )
            .setPriority(NotificationCompat.PRIORITY_HIGH)
            .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
            .setOnlyAlertOnce(true)
            .build()
        startForeground(1, notification)
    }
    inner class BindService : Binder() {
        val service: MediaService
            get() = this@MediaService
    }
    companion object {
        private var isExists: Boolean = false
        fun getIsExists() = isExists
        private var trackList = ArrayList<Track>()
        fun getMediaList(): ArrayList<Track> = trackList
    }
}